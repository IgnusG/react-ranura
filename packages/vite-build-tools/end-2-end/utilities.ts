import fs from "node:fs";
import path from "node:path";
import url from "node:url";

export const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

export function readFile(relativePath: string) {
	return fs.readFileSync(path.join(__dirname, relativePath)).toString();
}

export function cleanup() {
	fs.writeFileSync(
		path.join(__dirname, "fixtures/package.json"),
		JSON.stringify({ name: "fixtures" }, undefined, 2),
	);

	fs.rmSync(path.join(__dirname, "fixtures/dist"), { force: true, recursive: true });
}
