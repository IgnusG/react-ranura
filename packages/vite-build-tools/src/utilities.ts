import path from "node:path";
import { cwd } from "node:process";

import { ResolvedConfig, UserConfig } from "vite";

import { Entry, EntryOptions, PluginOptions } from "./types.js";

export function getPackageJSONPath({ config }: { config: ResolvedConfig | UserConfig }) {
	return path.join(config.root ?? cwd(), "package.json");
}

export function addRelativeDot(path: string) {
	if (path.startsWith("./")) return path;

	return `./${path}`;
}

export function removeRelativeDot(path: string) {
	if (!path.startsWith("./")) return path;

	return path.replace(/^\.\//, "");
}

/** Normalize input entry options and their paths into a standardized format
 *
 * sourcePath: absolute path to the entry's source file
 *
 * - outputPath:
 * 	   - if `entryOptions.outputPath` not provided an extensionless path based on the sourcePath relative to the root directory
 *     - if provided must be a relative extensionless path
 *
 * - exportPath:
 * 	   - if `entryOptions.exports.isMain` is true, "."
 *     - if `entryOptions.exports.exportPath` not provided the `./${outputPath}`
 *     - if provided must be a relative extensionless path
 */
export function createReduceEntryOptionsToEntries({
	config,
	options,
}: {
	config: UserConfig;
	options: PluginOptions;
}) {
	return function reduceEntryOptionsToEntries(
		result: Record<string, Entry>,
		entryOptions: EntryOptions,
		index: number,
	): Record<string, Entry> {
		const sourcePath = path.isAbsolute(entryOptions.sourcePath)
			? entryOptions.sourcePath
			: path.join(config.root ?? cwd(), entryOptions.sourcePath);

		const sourcePathChunks = path.parse(sourcePath);

		const outputPath =
			entryOptions.outputPath ??
			path.join(
				path.relative(
					path.join(config.root ?? cwd(), options.sourceRoot ?? ""),
					sourcePathChunks.dir,
				),
				sourcePathChunks.name,
			);

		const exportPath = entryOptions.exports?.exportPath ?? outputPath;

		const entry: Entry = {
			sourcePath,
			outputPath: removeRelativeDot(outputPath),
			exportPath: entryOptions.exports?.isMain ? "." : addRelativeDot(exportPath),
		};

		return {
			...result,
			[index.toString()]: entry,
		};
	};
}
