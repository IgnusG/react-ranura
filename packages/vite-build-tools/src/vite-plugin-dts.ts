import * as fs from "node:fs";
import * as path from "node:path";
import { cwd } from "node:process";

import { Plugin, ResolvedConfig } from "vite";

import { Entry, PackageJsonExports, PluginOptions } from "./types.js";
import { createReduceEntryOptionsToEntries, getPackageJSONPath } from "./utilities.js";

export type { PluginOptions };

type InputToOutput = Map<string, string>;

interface PackageJsonTypeExports {
	[exportPath: string]: {
		types: string;
	};
}

function createReduceEntriesToPackageExports({
	inputToOutput,
	buildConfig,
}: {
	inputToOutput: InputToOutput;
	buildConfig: ResolvedConfig["build"];
}) {
	function outputToExportPath(output: string) {
		return "./" + path.join(buildConfig.outDir, output);
	}

	return function reduceEntriesToPackageExports(
		result: PackageJsonTypeExports,
		entry: Entry,
	): PackageJsonTypeExports {
		const output = inputToOutput.get(entry.sourcePath);

		if (!output) {
			throw new Error(`Cannot find actual .dts output path for entry source "${entry.sourcePath}"`);
		}

		return {
			...result,

			[entry.exportPath]: {
				types: outputToExportPath(output),
			},
		};
	};
}

function createReduceExistingExportsEntriesToTypedPackageExports({
	entryTypeExports,
}: {
	entryTypeExports: Map<string, { types: string }>;
}) {
	return function reduceExistingExportsEntriesToTypedPackageExports(
		result: PackageJsonExports,
		[entryExportPath, entryExports]: [string, PackageJsonExports[string]],
	): PackageJsonExports {
		const entryTypeExport = entryTypeExports.get(entryExportPath);

		if (!entryTypeExport) {
			throw new Error(
				`Cannot find type definitions for export ${entryExportPath}. Searched exports ${exports.keys()}`,
			);
		}

		return {
			...result,
			[entryExportPath]: {
				...entryTypeExport,
				...entryExports,
			},
		};
	};
}

export default async function dtsPlugin(options: PluginOptions): Promise<Plugin> {
	let entries: Map<string, Entry>;
	let config: ResolvedConfig;

	const inputToOutput: InputToOutput = new Map();

	return {
		name: "vite:dts",
		config(userConfig) {
			entries = new Map(
				Object.entries(
					options.entries.reduce(
						createReduceEntryOptionsToEntries({ config: userConfig, options }),
						{},
					),
				),
			);
		},

		configResolved(userConfig) {
			config = userConfig;
		},

		generateBundle(outputOptions, bundle) {
			if (outputOptions.format !== options.formats[0]) return;

			for (const assetOrChunk of Object.values(bundle)) {
				if (assetOrChunk.type !== "chunk") continue;
				if (!assetOrChunk.facadeModuleId) continue;
				if (!assetOrChunk.isEntry) continue;

				const file = fs.readFileSync(assetOrChunk.facadeModuleId).toString();

				const hasDefaultExport = /^(export default |export \{[^}]+? as default\s*[,}])/m.test(file);

				const relativePath = path.relative(
					path.join(config.root ?? cwd(), assetOrChunk.fileName),
					assetOrChunk.facadeModuleId,
				);

				const relativeJsPath = relativePath.replace(/\.ts(x)?$/, ".js$1");

				const source =
					`export * from "${relativeJsPath}"` +
					(hasDefaultExport ? `\nexport {default} from "${relativeJsPath}"` : ``);

				const output = path.join(
					path.parse(assetOrChunk.fileName).dir,
					`${path.parse(assetOrChunk.fileName).name.replace(/\.(es|cjs)$/, "")}.d.ts`,
				);

				inputToOutput.set(assetOrChunk.facadeModuleId, output);

				this.emitFile({
					type: "asset",
					fileName: output,
					source,
				});
			}
		},

		closeBundle() {
			const packageText = fs.readFileSync(getPackageJSONPath({ config })).toString();
			const packageDetails = JSON.parse(packageText);

			const entryTypeExports = new Map(
				Object.entries(
					Array.from(entries.values()).reduce(
						createReduceEntriesToPackageExports({ inputToOutput, buildConfig: config.build }),
						{},
					),
				),
			);

			if (
				!packageDetails["#exports"] ||
				!packageDetails["#exports"].startsWith("Generated automatically")
			) {
				throw new Error(
					`Couldn't find the auto-generated marker #exports - add the vite-plugin-entries plugin before vite-plugin-dts (it must run after it)`,
				);
			}

			if (!options.silent)
				this.warn("Adding type definitions to the `exports` field in your package.json âœ…");

			packageDetails.exports = Array.from(
				Object.entries(packageDetails.exports as PackageJsonExports),
			).reduce(createReduceExistingExportsEntriesToTypedPackageExports({ entryTypeExports }), {});

			fs.writeFileSync(
				getPackageJSONPath({ config }),
				JSON.stringify(packageDetails, undefined, 4) + (packageText.endsWith("\n") ? "\n" : ""),
			);
		},
	};
}
