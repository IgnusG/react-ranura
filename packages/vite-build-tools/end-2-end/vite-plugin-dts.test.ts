import path from "node:path";

import { build, InlineConfig } from "vite";
import { afterAll, beforeEach, describe, expect, test } from "vitest";

import dtsPlugin from "../src/vite-plugin-dts.js";
import { PluginOptions } from "../src/vite-plugin-dts.js";
import entriesPlugin from "../src/vite-plugin-entries.js";
import { __dirname, cleanup, readFile } from "./utilities.js";

const defaultBuidConfig: InlineConfig = {
	root: path.join(__dirname, "fixtures"),
	build: {
		target: ["node18"],
		minify: false,
	},
};

describe("vite-plugin-entries", () => {
	beforeEach(() => {
		cleanup();
	});

	afterAll(() => {
		cleanup();
	});

	test("correctly creates the output build files for a simple es configuration", async () => {
		const buildConfig: PluginOptions = {
			formats: ["es"],
			entries: [{ sourcePath: path.join(__dirname, "fixtures/index.ts") }],
		};

		await build({
			...structuredClone(defaultBuidConfig),
			plugins: [entriesPlugin(buildConfig), dtsPlugin(buildConfig)],
		});

		expect(readFile("fixtures/dist/index-06b5571c.d.ts")).toMatchSnapshot();

		expect(JSON.parse(readFile("fixtures/package.json"))).toEqual(
			expect.objectContaining({
				exports: {
					"./index": expect.objectContaining({
						types: "./dist/index-06b5571c.d.ts",
					}),
				},
			}),
		);
	});
});
