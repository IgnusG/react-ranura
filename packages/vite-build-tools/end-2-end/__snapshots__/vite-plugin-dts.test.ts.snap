// Vitest Snapshot v1

exports[`vite-plugin-entries > correctly creates the output build files for a simple es configuration 1`] = `"export * from \\"../index.js\\""`;
